### PR TITLE
I18NPROD-322-add-method-to-format-value-for-gift-card

### PR DESCRIPTION
**Pull request by kllewellyn**
## What does this PR do?

Adds a method to `CommerceFormatters.java` that translates the "Value" string displayed when a product is a gift card and defaults to the user input variant name otherwise.

## Why is this important?

Currently there is a bug where the text above the gift card option drop-down displayed "Value" in English in all languages. This formatting method solves this bug.

## What high level changes did you make to the code to accomplish that goal?

* Added a method to `CommerceFormatters.java` that resolves the `displayName` based on the locale if a product is a gift card type otherwise it returns the `name` of the variant
* Changed the `variant-option-title` tag within `variants-select.html` to `{displayName}` with a default to `name` for non-gift card product types.

## Anything else?

This change also required adding the string `Value` to the `products-en-US.json` and `products-2.0-en-US.json` files within site-server to be translated into all languages. That PR can be found [here](https://github.com/sqsp/squarespace-v6/pull/11407) and will need to be merged and have translations returned before these changes are incorporated.

## How was this tested?

- [X] Made local changes to template-compiler and ran site-server with the local snapshot of template-compiler changes
- [X] Created a gift card locally and verified the translated text showed
- [X] Created a physical product with a unique variant name and verified that name displayed
- [X] Created a product with the suggested variant name and verified it displayed
- [x] Write additional tests for this functionality 

### Screenshots

Screen Shot 2021-07-22 at 10 17 21 AM
Screen Shot 2021-07-22 at 10 17 52 AM
Screen Shot 2021-07-22 at 11 09 05 AM

### CODE REVIEW

**Comment by phensley** [comment]
@kllewellyn  Thanks - can you add one or more new test cases for this? You can base your test on an existing example, like [f-variants-select-5.html](https://github.com/Squarespace/template-compiler/blob/master/core/src/test/resources/com/squarespace/template/plugins/platform/f-variants-select-5.html). 
Just name it with the next sequential integer `f-variants-select-6.html` and the test runner should automatically pick it up.

**Comment by kllewellyn** [comment]
> @kllewellyn Thanks - can you add one or more new test cases for this? You can base your test on an existing example, like [f-variants-select-5.html](https://github.com/Squarespace/template-compiler/blob/master/core/src/test/resources/com/squarespace/template/plugins/platform/f-variants-select-5.html).
> Just name it with the next sequential integer `f-variants-select-6.html` and the test runner should automatically pick it up.

Absolutely!

**Comment by xkb** [comment]
Nice work @kllewellyn! do we need equivalent changes in [Squarespace/template-engine](https://github.com/Squarespace/template-engine)?

**Comment by kllewellyn** [comment]
> Nice work @kllewellyn! do we need equivalent changes in [Squarespace/template-engine](https://github.com/Squarespace/template-engine)?

Yes! That is the intent - I only learned of template-engine yesterday so it is on the to-do list.

File: core/src/main/java/com/squarespace/template/plugins/platform/CommerceFormatters.java [hunk 1]
```diff
@@ -646,11 +646,24 @@ public void apply(Context ctx, Arguments args, Variables variables) throws CodeE
...
 
+    private static TextNode getDisplayName(Context ctx, JsonNode node) {
+      ProductType productType = CommerceUtils.getProductType(node);
+      // Gift Cards have variants forcibly named "Value" by default (as opposed to a merchant-defined variant name) and
+      // this string must be translated before being displayed on the front-end.
+      if (productType == ProductType.GIFT_CARD) {
+        String localizedDisplayName = ctx.resolve(Constants.GIFT_CARD_VALUE_DISPLAY_TEXT).asText();
+        return new TextNode(StringUtils.defaultIfEmpty(localizedDisplayName, "Value"));
+      } else {
+        return new TextNode(StringUtils.defaultString("{name}"));
```
Comment (jpoo90) [inline core/src/main/java/com/squarespace/template/plugins/platform/CommerceFormatters.java:663]: Could `name` be defined for gift cards?

File: core/src/main/resources/com/squarespace/template/plugins/platform/variants-select.html [hunk 2]
```diff
@@ -6,7 +6,7 @@
 
   <div class="product-variants" data-item-id="{item.id}" data-variants="{@variants|json|htmltag}"{.if @isSubscribable} data-is-subscribable="true" data-subscription-plan="{@subscriptionPlan|json|htmltag}"{.end} data-animation-role="content">{.repeated section options}
     <div class="variant-option">
-    <div class="variant-option-title">{name|htmltag}: </div>
+    <div class="variant-option-title">{displayName|message name:name|htmltag}: </div>
```
Comment (jpoo90) [inline core/src/main/resources/com/squarespace/template/plugins/platform/variants-select.html:9]: Isn't `name` duplicated here ?

File: core/src/test/resources/com/squarespace/template/plugins/platform/f-variants-select-4.html [hunk 3]
```diff
@@ -1,12 +1,12 @@
 :JSON
 {
   "localizedStrings": {
-    "giftCardVariantSelectText": "Get foobar"
+    "productVariantSelectText": "Get foobar"
```
Comment (jpoo90) [inline core/src/test/resources/com/squarespace/template/plugins/platform/f-variants-select-4.html:4]: Is this change intended ?

File: core/src/main/java/com/squarespace/template/plugins/platform/CommerceFormatters.java (hunk 1 above)
Comment (kllewellyn) [reply core/src/main/java/com/squarespace/template/plugins/platform/CommerceFormatters.java:663]: It is - it will be "Value" for a gift card every time.

File: core/src/main/resources/com/squarespace/template/plugins/platform/variants-select.html (hunk 2 above)
Comment (kllewellyn) [reply core/src/main/resources/com/squarespace/template/plugins/platform/variants-select.html:9]: It is substituting in the name for the field rather than the displayName

File: core/src/test/resources/com/squarespace/template/plugins/platform/f-variants-select-4.html (hunk 3 above)
Comment (kllewellyn) [reply core/src/test/resources/com/squarespace/template/plugins/platform/f-variants-select-4.html:4]: Yes, but it depends. The intention of this test originally appeared to be testing gift card behavior. See product type = 4 but after the changes I have made, the titles of the variants will not take the {name} it will instead take the translated value of "Value" so this test would change quite a bit if I were to modify it for a gift card. Instead, I switched it to a physical product test but can revert it back / make modifications to test it as a gift card.

**Review by phensley — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/squarespace/template/plugins/platform/CommerceFormatters.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (kllewellyn) [reply core/src/main/java/com/squarespace/template/plugins/platform/CommerceFormatters.java]: Good point - this has been updated!

**Review by phensley — APPROVED** [review]

**Comment by phensley** [comment]
@kllewellyn Ready to merge?

**Comment by phensley** [comment]
@kllewellyn This is released in `2.8.19`
